proposals/0008-package-manager-testing.md

```diff
@@ -0,0 +1,194 @@
+# Swift Testing
+
+* Authors:
+  [Max Howell](https://github.com/mxcl),
+  [Daniel Dunbar](https://github.com/ddunbar),
+  [Mattt Thompson](https://github.com/mattt)
+* Status: **Review**
+* Review Manager: Rick Ballard
+
+## Introduction
+
+Testing is an essential part of modern software development.
+Tight integration of testing into the Swift Package Manager
+will help ensure a stable and reliable packaging ecosystem.
+
+## Proposed Solution
+
+We propose to extend our conventional package directory layout
+to accomodate test modules.
+Any subdirectory of the package root directory named "Tests"
+or any subdirectory of an existing module directory named "Tests"
+will comprise a test module.
+For example:
+
+    Package
+    ├── Sources
+    │   └── Foo
+    │       └──Foo.swift
+    └── Tests
+        └── Foo
+            └── Test.swift
+
+Or:
+
+    Package
+    └── Sources
+        ├── Foo.swift
+        └── Tests
+            └── Test.swift
+
+Or, for simpler projects:
+
+    Package
+    ├── Sources
+    │   └── Foo.swift
+    └── Tests
+        └── TestFoo.swift
+        
+The filename: `TestFoo.swift` is arbitrary.
+
+In the examples above
+a test case is created for the module `Foo`
+based on the sources in the relevant subdirectories.
+
+Additionally we will support directories called `FooTests`.
+This layout style is prevalent in existing open source projects
+and supporting it will minimize vexation for their authors.
+However in the interest of consistency and the corresponding 
+reduction of cognitive-load when examining new Swift packages
+we will not recommend this layout. For example:
+
+    Package
+    └── Sources
+    │   └── Foo.swift
+    └── FooTests
+        └── Test.swift
+
+Additionally, we propose that building a module
+also builds that module's corresponding tests.
+Although this would result in slightly increased build times,
+we believe that tests are important enough to justify this
+(one might even consider slow building tests to be a code smell).
+We would prefer to go even further by executing the tests
+each time a module is built as well,
+but we understand that this would impede debug cycles.
+
+As an exception, when packages are built in release mode we will
+not build tests because for release builds we should not enable
+testability. However, considering the need for release-mode testing
+this will be a future direction.
+
+### Command-Line Interface
+
+We propose the following syntax to execute tests:
+
+    $ swift build --test
+
+Or:
+
+    $ swift build -t
+
+In the future, we may choose to promote the `--test` option
+to be a subcommand of the `swift` command itself:
+
+    $ swift test
+
+However, any such decision would warrant extensive design consideration,
+so as to avoid polluting or crowding the command-line interface.
+Should there be sufficient demand and justification for it, though,
+it would be straightforward to add this functionality.
```

Comment:
Very much agree in terms of command-line arguments being used by test runners, and in terms of the **necessity** of custom test reporters.

Still, I believe Swift's `Process.environment` may be able to suffice for now, and would allow us to decouple that important discussion from this proposal.